database/src/table.rs

```diff
@@ -1,181 +1,49 @@
-//! [`DupTable`] are just a trait used to define that they support DUPSORT|DUPFIXED operation (as of now we don't know the equivalent for HSE).
-//! All tables are defined with docs explaining its purpose, what types are the key and data.
-//! For more details please look at Cuprate's book : <link to cuprate book>
-
-use crate::{
-    encoding::Compat,
-    types::{
-        /*OutTx,*/ AltBlock, BlockMetadata, /*RctOutkey,*/ OutputMetadata,
-        TransactionPruned, TxIndex, /*OutAmountIdx,*/ /*KeyImage,*/ TxOutputIdx,
-    },
-};
-use bincode::{de::Decode, enc::Encode};
-use monero::{blockdata::transaction::KeyImage, Block, Hash};
-
-/// A trait implementing a table interaction for the database. It is implemented to an empty struct to specify the name and table's associated types. These associated
-/// types are used to simplify deserialization process.
-pub trait Table: Send + Sync + 'static + Clone {
-    // name of the table
-    const TABLE_NAME: &'static str;
-
-    // Definition of a key & value types of the database
-    type Key: Encode + Decode;
-    type Value: Encode + Decode;
-}
-
-/// A trait implementing a table with duplicated data support.
-pub trait DupTable: Table {
-    // Subkey of the table (prefix of the data)
-    type SubKey: Encode + Decode;
-}
-
-/// This declarative macro declare a new empty struct and impl the specified name, and corresponding types.
-macro_rules! impl_table {
-	( $(#[$docs:meta])* $table:ident , $key:ty , $value:ty ) => {
-        #[derive(Clone)]
-		$(#[$docs])*
-		pub(crate) struct $table;
-
-   		impl Table for $table {
-	 		const TABLE_NAME: &'static str = stringify!($table);
-        	type Key = $key;
-            type Value = $value;
+//! Database table abstraction; `trait Table`.
+
+//---------------------------------------------------------------------------------------------------- Import
+use crate::{key::Key, pod::Pod};
+
+//---------------------------------------------------------------------------------------------------- Table
+/// Database table metadata.
+///
+/// Purely compile time information for database tables.
+/// Not really an accurate name for `K/V` database but
+/// this represents the metadata of a `K/V` storing object.
+pub trait Table {
+    // TODO:
+    //
+    // Add K/V comparison `type`s that define
+    // how this table will be stored.
+    //
+    // type KeyComparator: fn(&Self::Key, &Self::Key) -> Ordering;
+    // type ValueComparator: fn(&Self::Value, &Self::Value) -> Ordering;
+
+    /// Name of the database table.
+    const NAME: &'static str;
+
+    /// Whether the table's values are all the same size or not.
+    const CONSTANT_SIZE: bool;
+
+    // TODO: fix this sanakirja bound.
+    cfg_if::cfg_if! {
+        if #[cfg(all(feature = "sanakirja", not(feature = "heed")))] {
+            /// Primary key type.
+            type Key: Key + Pod + sanakirja::Storable;
```

Comment:
The `Pod` and `Storable` bound should be on the `Primary` and `Secondary` types of the `Key` trait

Reply:
This means `DupKey` needs to implement `Pod`. How should it be implemented?

Reply:
https://github.com/Cuprate/cuprate/pull/61